member/src/main/scala/com/improving/app/member/domain/Member.scala

```diff
@@ -56,220 +68,409 @@ object Member extends StrictLogging {
+    }
   }
-  def terminateMember(
-      memberId: MemberId,
-      terminatingMember: MemberId,
-      replyTo: ActorRef[StatusReply[MemberResponse]]
+
+  private def terminateMemberLogic(
+    meta: MemberMetaInfo,
+    terminateMemberCommand: TerminateMember,
+    replyTo: ActorRef[StatusReply[MemberResponse]]
   ): ReplyEffect[MemberEvent, MemberState] = {
-    val event = MemberTerminated(Some(memberId), Some(terminatingMember), Some(Timestamp(Instant.now(clock))))
+    val newMeta = meta.copy(
+      lastModifiedBy = terminateMemberCommand.terminatingMember,
+      lastModifiedOn = Some(Timestamp(Instant.now(clock)))
+    )
+    val event = MemberTerminated(terminateMemberCommand.memberId, Some(newMeta))
     Effect.persist(event).thenReply(replyTo) { _ => StatusReply.Success(MemberEventResponse(event)) }
-
   }
 
-  def updateMemberInfo(
-      memberId: MemberId,
-      memberInfo: MemberInfo,
-      actingMember: MemberId,
-      replyTo: ActorRef[StatusReply[MemberResponse]]
-  ): ReplyEffect[MemberEvent, MemberState] =
-    MemberValidation.validateMemberInfo(memberInfo) match {
-      case Valid(memberInfo) =>
-        val event =
-          MemberInfoUpdated(Some(memberId), Some(memberInfo), Some(actingMember), Some(Timestamp(Instant.now(clock))))
-        Effect.persist(event).thenReply(replyTo) { _ => StatusReply.Success(MemberEventResponse(event)) }
-      case Invalid(errors) =>
-        Effect.reply(replyTo) {
-          StatusReply.Error(
-            s"Invalid Member Info: $memberInfo with errors: ${errors.map { _.errorMessage }.toList.mkString(",")}"
-          )
+  private def terminateMember(
+    state: MemberState,
+    terminateMemberCommand: TerminateMember,
+    replyTo: ActorRef[StatusReply[MemberResponse]]
+  ): ReplyEffect[MemberEvent, MemberState] = {
+    MemberValidation.validateTerminateMemberCommand(terminateMemberCommand) match {
+      case Valid(_) =>
+        state match {
+          case _: DraftMemberState => useErrorStatusReply(replyTo, s"Member has not yet been registered.")
+          case RegisteredMemberState(_, meta) => terminateMemberLogic(meta, terminateMemberCommand, replyTo)
+          case _: TerminatedMemberState => useTerminatedStateErrorStatusReply(replyTo)
         }
+      case Invalid(errors) =>
+        useCommandValidationErrorStatusReply(replyTo, errors, "TerminateMember")
     }
+  }
 
-  def getMemberInfo(
-      memberId: MemberId,
-      state: MemberState,
-      value: ActorRef[StatusReply[MemberResponse]]
+  private def editMemberInfoLogic(
+    info: MemberInfo,
+    metaInfo: MemberMetaInfo,
+    editMemberInfoCommand: EditMemberInfo,
+    replyTo: ActorRef[StatusReply[MemberResponse]]
   ): ReplyEffect[MemberEvent, MemberState] = {
-    Effect.reply(value) {
-      StatusReply.Success(
-        MemberData(Some(memberId), state.memberInfo, state.memberMetaInfo)
+    if (metaInfo.createdBy.isEmpty) {
```

Comment:
Another spot where the presence of the createdBy is standing in for the actor state instead of using the actor state explicitly (which we can do in a way to get a compile-time guarantee, not to sound like a broken record)